datafusion/src/physical_plan/file_format/parquet.rs

```diff
@@ -446,6 +471,62 @@ fn build_row_group_predicate(
     }
 }
 
+fn read_partition_no_file_columns(
```

Comment:
I haven't looked carefully at this code, but it seems like this has a lot of duplication with `read_partition` -- I wonder if it is possible somehow to unify the code paths, perhaps by checking `projection.is_empty()` in `read_partition`

Reply:
You don't need to open all row group to get the number of rows in parquet, you have all this information in the footer. You should use `file_reader.metadata()` here. Once you do that, you can spare yourself the limit logic that is pretty verbose. You just iterate through all row groups in the metadata to count all the rows in the file, that's very cheap because the data structure is loaded to memory when the footer is parsed. This should simplify greatly this code path, then we can re-evaluate if we need to merge it with the one above or not 😉.

Reply:
Missed the file level metadata. I've reworked the function to use that, and it has simplified it quite a bit.